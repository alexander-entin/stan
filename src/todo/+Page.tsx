import { TodoList } from './TodoList'
import { NewTodoForm } from './NewTodoForm'
import { TodoToolbar } from './TodoToolbar'

export default function Page() {
	console.log('todo.render')
	return (<>
		<h1>To-do List</h1>
		<NewTodoForm />
		<TodoToolbar />
		<TodoList />
	</>)
}
