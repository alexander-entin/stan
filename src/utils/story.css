.diff {
	/* customization for the top-level <div> wrapper that contains the diff */
	font-family: ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
	font-size: 1em;
	white-space: pre;
}

.deletion {
	/* customization for deleted lines (<div>) */
	color: #ff0000;
}

.addition {
	/* customization for added lines (<div>) */
	color: #00dd00;
}

.unchanged {
	/* customization for unchanged lines (<div>) */
	color: gray;
}
