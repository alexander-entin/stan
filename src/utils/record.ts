import { $global, $event, snapshot, subscribe } from '../utils/stan'

if (typeof window !== 'undefined') {
	setTimeout(() => {
		let initial
		let events = []
		window.record = () => {
			if (events.length) {
				console.log({
					initial,
					events,
				})
			}
			initial = snapshot($global)
		}
		record()
		subscribe($event, () => {
			events.push(snapshot($event))
		})
	}, 1)
}
