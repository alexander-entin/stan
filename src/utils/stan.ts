import { proxy, subscribe, snapshot, useSnapshot, ref } from 'valtio'
import { subscribeKey, derive, watch, devtools } from 'valtio/utils'
import { nanoid } from 'nanoid'
import { path } from 'rambda'

export { subscribe, subscribeKey, snapshot, watch, derive, ref }
export const useStan = useSnapshot

export type Entity = {
	id: string,
	[key: string]: unknown,
}

export type SyncState = {
	status?: string,
	error?: unknown,
	pull: {
		[reqId: string]: {
			query: unknown,
			at: number,
			doneAt?: number,
			error?: unknown,
		},
	},
	push: {
		[reqId: string]: {
			method: string,
			body: Entity,
			prev: Entity,
			at: number,
			doneAt?: number,
			error?: unknown,
		},
	}
}

export const $global = proxy({
	sync: {} as Record<string, SyncState>,
})
export const $sync = $global.sync
export const $event = proxy({
	id: nanoid(),
	at: Date.now(),
	type: 'init',
	data: [] as any[],
})

type Reactor = (...payload: any[]) => void
const reactors = {} as { [event: string]: { [module: string]: Reactor } }
export function createReactor<F extends Reactor>(module: string, event: string, fn: F) {
	let rs = reactors[event] ?? {}
	reactors[event] = rs
	let id = module ?? Object.keys(rs).length
	rs[id] = fn
	return dispatch({ type: event }) as F
}

export type EventMeta = {
	type: string,
	id: string,
	at: number,
}
export type Event = EventMeta & { data?: any[] }

let initial
export const dispatch = ({ id, at, type }: Partial<EventMeta>) => (...payload: unknown[]) => {
	if (import.meta.env.DEV && typeof window === 'object') {
		initial ||= snapshot($global)
	}
	$event.id = id ?? nanoid()
	$event.at = at ?? Date.now()
	$event.type = type!
	$event.data = payload
	console.warn(type, JSON.stringify(payload))
	Object.values(reactors[type!]).forEach(f => {
		// console.warn(type, JSON.stringify(payload), { f })
		f(...payload)
	})
	// console.warn(snapshot($global))
}

function isVirtualProp(o, k) {
	let d = Object.getOwnPropertyDescriptor(o, k) || {}
	return d.get || d.set
}

function isDataProp(o, k) {
	return !isVirtualProp(o, k) && typeof o[k] !== 'function'
}

export function createStan<T extends object>(name: string, stan: T): T {
	Object.keys(stan).forEach(k => {
		if (!isVirtualProp(stan, k)
			&& typeof stan[k] === 'function'
			&& k.startsWith('on')
			&& k[2] === k[2]?.toUpperCase()
		) {
			let event = `${name}.${k}`
			stan[k] = createReactor(name, event, stan[k])
		}
	})
	$global[name] = stan
	return $global[name] as T
}

function dive(o, f, p = [] as string[]) {
	Object.keys(o).forEach(k => {
		if (!isDataProp(o, k)) return
		f(o, k, p)
		if (typeof o[k] === 'object') {
			dive(o[k], f, [...p, k])
		}
	})
}

export function reset($stan, snap = initial) {
	if (!snap) return
	dive(snap, (s, k, p) => {
		const $ = path(p, $stan)
		if (!isDataProp($, k)) return
		if (typeof s[k] !== 'object' || typeof $[k] !== 'object') {
			$[k] = structuredClone(s[k])
		}
	})
	dive($stan, ($, k, p) => {
		const s = path(p, snap)
		if (!(k in s)) {
			delete $[k]
		}
	})
}

export function integrated() {
	return import.meta.env.MODE !== 'test' && import.meta.env.MODE !== 'story'
}

export type Paths = string | string[] | string[][]
export type SubscribeCallback = (ops: unknown[][]) => void
export function subscribePaths(proxy: object, paths: Paths, fn: SubscribeCallback, ...rest) {
	if (typeof paths === 'string') paths = paths.split(',')
	paths = paths.map(x => typeof x === 'string' ? x.split('.') : x)
	return subscribe(proxy, ops => {
		ops = ops.filter(([_, path]) =>
			(paths as string[][]).find(p => p.every((x, i) => path[i] === x))
		)
		ops.length && fn(ops)
	}, ...rest)
}

if (import.meta.env.DEV && typeof window === 'object') {
	Object.assign(window, { $global, $event, proxy, subscribe, watch, derive, snapshot, reset })
	devtools($global, { name: '$global', enabled: true })

	let snap
	let events = [] as Event[]
	function story() {
		const indent = snap ? '\t' : ''
		let json = ['[', ...events.map(({ type, data = [] }) => {
			let rest = [data]
			if (data.length === 0) {
				rest = []
			} else if (data.length === 1 && !Array.isArray(data[0])) {
				rest = data
			}
			return `\t${JSON.stringify([type, ...rest])},`
		}), ']'].join(`\n${indent}`)
		if (snap) {
			let init = JSON.stringify(snap, null, '\t')
			init = init.split('\n').map((r, i) => `${i ? '\t' : ''}${r}`).join('\n')
			json = `{\n\tinitial: ${init},\n\tevents: ${json}\n}`
		}
		const val = snap ? { initial: snap, events } : events
		events = []
		snap = snapshot($global)
		dive(snap, (s, k, p) => {
			const $ = path(p, $global)
			if (!isDataProp($, k)) {
				delete s[k]
			}
		})
		console.log(json)
		return val
	}
	subscribe($event, () => {
		events.push(snapshot($event) as Event)
	})
	window['story'] = story
}
