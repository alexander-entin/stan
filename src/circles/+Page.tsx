import LocalStorage from "../drivers/localStorage"
import { createStan, useStan } from "../utils/stan"

type Point = {
	x: number,
	y: number,
}

const $circles = createStan('circles', {
	list: [] as Point[],
	size: 0,
	onAdd(x, y) {
		$circles.list[$circles.size++] = { x, y }
		$circles.list.length = $circles.size
	},
	onUndo() {
		if ($circles.size) {
			$circles.size--
		}
	},
	onRedo() {
		if ($circles.size < $circles.list.length) {
			$circles.size++
		}
	}
})

// LocalStorage.sync('circles', $circles)

export function Page() {
	const cs = useStan($circles)
	return <>
		<div className="absolute left-0 top-0">
			<button
				className="btn"
				disabled={!cs.size}
				onClick={cs.onUndo}
			>
				<span className="i-mdi-undo text-xl" />
			</button>
			<button
				className="btn"
				disabled={cs.size === cs.list.length}
				onClick={cs.onRedo}
			>
				<span className="i-mdi-redo text-xl" />
			</button>
		</div>
		<div
			className="h-100vh"
			onClick={e => cs.onAdd(e.clientX, e.clientY)}
		>
			{cs.list.slice(0, cs.size).map((c, i) =>
				<div
					className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
					style={{ top: c.y, left: c.x }}
					key={i}
				/>
			)}
		</div>
	</>
}