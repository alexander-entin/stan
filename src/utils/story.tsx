import '@unocss/reset/tailwind.css'
import 'virtual:uno.css'
import { FC, useMemo } from 'react'
import deepRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import { deepmergeInto } from 'deepmerge-ts'
import { useLocalStorage } from 'react-use'
import JsonView from 'react18-json-view'
import 'react18-json-view/src/style.css'
import 'react18-json-view/src/dark.css'
import { JsonDiffComponent } from 'json-diff-react'
import { diff } from 'deep-object-diff'

import './story.css'
import { Event, $global, snapshot, dispatch, replay } from './stan'

export function normalize(story) {
	if (Array.isArray(story)) {
		story = { events: story }
	}
	let t = new Date('2020-02-02').valueOf()
	story.events = story.events?.map(event => {
		let id, at, type, data
		if (Array.isArray(event)) {
			[type, data] = event
		} else {
			({ id, at, type, data } = event)
		}
		id ||= `#${t}`
		at ||= t++
		data = Array.isArray(data) ? data : data === undefined ? [] : [data]
		return { id, at, type, data }
	}) || []
	return story
}

function JsonDiff({ a, b }) {
	const [expanded, setExpanded] = useLocalStorage('JsonDiff.expanded', false)
	return (
		<div
			className="h-full"
			onClick={() => setExpanded(!expanded)}
		>
			<JsonDiffComponent
				jsonA={a}
				jsonB={b}
				jsonDiffOptions={{
					full: expanded,
					maxElisions: 2,
					renderElision: n => n > 1 ? `... (${n})` : '...',
				}}
			/>
		</div>
	)
}

const emptyArr = []
export function Story({ initial, events = emptyArr as Event[], children }) {
	const event = events[events.length - 1]
	const [tab, setTab] = useLocalStorage('story.tab', 'UI')
	const states = useMemo(() => replay(events, initial), [events, initial])
	let [prev, curr] = states.slice(-2)
	const tabs = {
		'Event': () => event && <JsonView src={event} />,
		'State': () => event ? <JsonDiff a={prev} b={curr} /> : <JsonView src={prev} />,
		'UI': () => children,
	}
	return <>
		{(tabs[tab!] ?? tabs.UI)()}
		<div className="tabs fixed bottom-0 mx-auto left-50% -translate-x-50% z-999">
			{Object.keys(tabs).map(name =>
				<a
					className={`tab tab-bordered ${name === tab && 'tab-active'}`}
					onClick={() => setTab(name)}
					key={name}
				>
					{name}
				</a>
			)}
		</div>
	</>
}

export function tell(story, Component: FC) {
	const { initial, events } = normalize(story)
	const fs = {
		'0. initial': <Story initial={initial}><Component /></Story>
	}
	events.forEach(({ type }, idx) => {
		fs[`${idx + 1}. ${type}`] = (
			<Story initial={initial} events={events.slice(0, idx + 1)}>
				<Component />
			</Story>
		)
	})
	return fs
}

export type DeepShallow = { [method in 'deep' | 'shallow']?: Record<string, FC> }
export function test(story, { it })
export function test(story, component: FC, { it })
export function test(story, components: DeepShallow, { it })
export function test(story, ...rest) {
	const { it } = rest.pop()
	const components = rest[0]
	let { initial = {}, events } = normalize(story)
	deepmergeInto($global, initial)
	let snap = snapshot($global)
	let { deep, shallow } = components || {}
	if (!deep && !shallow && components) {
		deep = { '': components }
	}
	events.forEach((event, idx) => {
		const { id, at, type, data } = event
		dispatch({ id, at, type })(...(Array.isArray(data) ? data : [data]))
		const curr = snapshot($global)
		const d = diff(snap, curr)
		snap = curr
		const prefix = `${idx + 1}. ${type} >`
		it(`${prefix} $stan changes`, ({ expect }) => {
			expect(d).toMatchSnapshot()
		})
		Object.entries(shallow ?? {}).forEach(([name, Component]) => {
			const shallowRenderer = new ShallowRenderer()
			const shallow = shallowRenderer.render(<Component />)
			it(`${prefix} ${name} (shallow)`, ({ expect }) => {
				expect(shallow).toMatchSnapshot()
			})
		})
		Object.entries(deep ?? {}).forEach(([name, Component]) => {
			const deep = deepRenderer.create(<Component />).toJSON()
			it(`${prefix} ${name} (deep)`, ({ expect }) => {
				expect(deep).toMatchSnapshot()
			})
		})
	})
}
