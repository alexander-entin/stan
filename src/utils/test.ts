import { it } from 'vitest'

import { dispatch, $global, snapshot } from '../utils/stan'

export function test($stan, cases) {
	Object.entries(cases).forEach(([name, spec]) => {
		it(name, ({ expect }) => {
			spec.events.forEach(({ data, ...meta }) =>
				dispatch(meta)(...(Array.isArray(data) ? data : [data]))
			)
			const snap = snapshot($global)
			expect(snap).toMatchSnapshot()
		})
	})
}