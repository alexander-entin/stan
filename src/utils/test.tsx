import { describe, it } from 'vitest'
import { diff } from 'deep-object-diff'
import deepRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'

import { dispatch, $global, snapshot } from './stan'

export function test(cases, defaultComponents) {
	Object.entries(cases).forEach(([name, { events, components }]) => {
		components = components || defaultComponents
		describe(name, () => {
			let t = new Date('2020-02-02').valueOf()
			let snap = snapshot($global)
			events.forEach((event, idx) => {
				const { id = `#${t}`, at = t, type, data } = event
				dispatch({ id, at, type })(...(Array.isArray(data) ? data : [data]))
				const curr = snapshot($global)
				const d = diff(snap, curr)
				delete d.event
				snap = curr
				it(`${idx}. ${type} > $stan changes`, ({ expect }) => {
					expect(d).toMatchSnapshot()
				})
				const cs = typeof components === 'function' ? components(event, idx) : components
				if (cs) {
					Object.entries(cs).forEach(([name, Component]) => {
						const component = <Component />
						const shallowRenderer = new ShallowRenderer()
						const shallow = shallowRenderer.render(component)
						it(`${idx}. ${type} > ${name} (shallow)`, ({ expect }) => {
							expect(shallow).toMatchSnapshot()
						})
						const deep = deepRenderer.create(component).toJSON()
						it(`${idx}. ${type} > ${name} (deep)`, ({ expect }) => {
							expect(deep).toMatchSnapshot()
						})
					})
				}
				t++
			})
		})
	})
}