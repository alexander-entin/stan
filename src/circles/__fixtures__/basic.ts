import { tell, test } from '/src/utils/story'
import { Page } from '../+Page'

const story = [
	["circles.onAdd", [190, 115]],
	["circles.onAdd", [283, 115]],
	["circles.onAdd", [362, 113]],
	["circles.onUndo"],
	["circles.onUndo"],
	["circles.onUndo"],
	["circles.onRedo"],
	["circles.onAdd", [190, 57]],
	["circles.onAdd", [190, 7]],
	["circles.onUndo"],
	["circles.onUndo"],
	["circles.onUndo"],
	["circles.onRedo"],
	["circles.onRedo"],
	["circles.onRedo"],
]

export default tell(story, Page)

if (import.meta.vitest) {
	test(story, Page, import.meta.vitest)
}
