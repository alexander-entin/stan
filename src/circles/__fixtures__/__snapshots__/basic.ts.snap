// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`1. circles.onAdd >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
  </div>,
]
`;

exports[`1. circles.onAdd > $stan changes 1`] = `
{
  "circles": {
    "list": {
      "0": {
        "x": 71,
        "y": 101,
      },
    },
    "size": 1,
  },
}
`;

exports[`2. circles.onAdd >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 134,
          "top": 101,
        }
      }
    />
  </div>,
]
`;

exports[`2. circles.onAdd > $stan changes 1`] = `
{
  "circles": {
    "list": {
      "1": {
        "x": 134,
        "y": 101,
      },
    },
    "size": 2,
  },
}
`;

exports[`3. circles.onAdd >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 134,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 198,
          "top": 99,
        }
      }
    />
  </div>,
]
`;

exports[`3. circles.onAdd > $stan changes 1`] = `
{
  "circles": {
    "list": {
      "2": {
        "x": 198,
        "y": 99,
      },
    },
    "size": 3,
  },
}
`;

exports[`4. circles.onUndo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 134,
          "top": 101,
        }
      }
    />
  </div>,
]
`;

exports[`4. circles.onUndo > $stan changes 1`] = `
{
  "circles": {
    "size": 2,
  },
}
`;

exports[`5. circles.onUndo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
  </div>,
]
`;

exports[`5. circles.onUndo > $stan changes 1`] = `
{
  "circles": {
    "size": 1,
  },
}
`;

exports[`6. circles.onUndo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  />,
]
`;

exports[`6. circles.onUndo > $stan changes 1`] = `
{
  "circles": {
    "size": 0,
  },
}
`;

exports[`7. circles.onRedo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
  </div>,
]
`;

exports[`7. circles.onRedo > $stan changes 1`] = `
{
  "circles": {
    "size": 1,
  },
}
`;

exports[`8. circles.onAdd >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 115,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 57,
        }
      }
    />
  </div>,
]
`;

exports[`8. circles.onAdd > $stan changes 1`] = `
{
  "circles": {
    "list": {
      "1": {
        "x": 190,
        "y": 57,
      },
      "2": undefined,
    },
    "size": 2,
  },
}
`;

exports[`8. circles.onRedo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 134,
          "top": 101,
        }
      }
    />
  </div>,
]
`;

exports[`8. circles.onRedo > $stan changes 1`] = `
{
  "circles": {
    "size": 2,
  },
}
`;

exports[`9. circles.onAdd >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 134,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 101,
          "top": 148,
        }
      }
    />
  </div>,
]
`;

exports[`9. circles.onAdd > $stan changes 1`] = `
{
  "circles": {
    "list": {
      "2": {
        "x": 101,
        "y": 148,
      },
    },
    "size": 3,
  },
}
`;

exports[`10. circles.onUndo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 134,
          "top": 101,
        }
      }
    />
  </div>,
]
`;

exports[`10. circles.onUndo > $stan changes 1`] = `
{
  "circles": {
    "size": 2,
  },
}
`;

exports[`11. circles.onRedo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 71,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 134,
          "top": 101,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 101,
          "top": 148,
        }
      }
    />
  </div>,
]
`;

exports[`11. circles.onRedo > $stan changes 1`] = `
{
  "circles": {
    "size": 3,
  },
}
`;

exports[`11. circles.onUndo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 115,
        }
      }
    />
  </div>,
]
`;

exports[`11. circles.onUndo > $stan changes 1`] = `
{
  "circles": {
    "size": 1,
  },
}
`;

exports[`12. circles.onUndo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  />,
]
`;

exports[`12. circles.onUndo > $stan changes 1`] = `
{
  "circles": {
    "size": 0,
  },
}
`;

exports[`13. circles.onRedo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 115,
        }
      }
    />
  </div>,
]
`;

exports[`13. circles.onRedo > $stan changes 1`] = `
{
  "circles": {
    "size": 1,
  },
}
`;

exports[`14. circles.onRedo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 115,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 57,
        }
      }
    />
  </div>,
]
`;

exports[`14. circles.onRedo > $stan changes 1`] = `
{
  "circles": {
    "size": 2,
  },
}
`;

exports[`15. circles.onRedo >  (deep) 1`] = `
[
  <div
    className="absolute left-0 top-0"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="i-mdi-undo text-xl"
      />
    </button>
    <button
      className="btn"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="i-mdi-redo text-xl"
      />
    </button>
  </div>,
  <div
    className="h-100vh"
    onClick={[Function]}
  >
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 115,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 57,
        }
      }
    />
    <div
      className="h-4 w-4 rounded-full absolute -m-2 border-white border-solid border-2"
      style={
        {
          "left": 190,
          "top": 7,
        }
      }
    />
  </div>,
]
`;

exports[`15. circles.onRedo > $stan changes 1`] = `
{
  "circles": {
    "size": 3,
  },
}
`;
